File: apps/agor-android/app/src/main/kotlin/live/agor/app/models/InputRequest.kt
package live.agor.app.models

import androidx.compose.runtime.Immutable
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

@Serializable
enum class InputRequestKind {
    @SerialName("free_text") FREE_TEXT,
    @SerialName("single_choice") SINGLE_CHOICE,
    @SerialName("multi_choice") MULTI_CHOICE,
}

@Serializable
enum class InputRequestStatus {
    @SerialName("pending") PENDING,
    @SerialName("answered") ANSWERED,
    @SerialName("cancelled") CANCELLED,
    @SerialName("timed_out") TIMED_OUT,
}

@Serializable
@Immutable
data class InputRequestOption(
    val label: String,
    val description: String = "",
    val markdown: String? = null,
)

@Serializable(with = InputRequestQuestionSerializer::class)
@Immutable
data class InputRequestQuestion(
    val question: String,
    val header: String? = null,
    val kind: InputRequestKind = InputRequestKind.FREE_TEXT,
    val options: List<InputRequestOption>? = null,
    val multiSelect: Boolean? = null,
)

@Serializable
@Immutable
data class InputRequestContent(
    @SerialName("request_id") val requestId: String = "",
    @SerialName("input_request_id") val legacyInputRequestId: String? = null,
    @SerialName("task_id") val taskId: String? = null,
    val questions: List<InputRequestQuestion> = emptyList(),
    val status: InputRequestStatus = InputRequestStatus.PENDING,
    val answers: Map<String, String>? = null,
    @SerialName("answered_at") val answeredAt: String? = null,
    @SerialName("requested_at") val requestedAt: String? = null,
    val context: String? = null,
) {
    val inputRequestId: String
        get() = requestId.ifBlank { legacyInputRequestId.orEmpty() }
}

object InputRequestQuestionSerializer : KSerializer<InputRequestQuestion> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("InputRequestQuestion")

    override fun deserialize(decoder: Decoder): InputRequestQuestion {
        val jd = decoder as? JsonDecoder ?: error("InputRequestQuestion requires JSON")
        val obj = jd.decodeJsonElement().jsonObject
        val kind = obj["kind"]?.let {
            jd.json.decodeFromJsonElement(InputRequestKind.serializer(), it)
        } ?: InputRequestKind.FREE_TEXT
        val options = (obj["options"] as? kotlinx.serialization.json.JsonArray)?.mapNotNull { option ->
            when (option) {
                is JsonPrimitive -> InputRequestOption(option.contentOrNull.orEmpty())
                is JsonObject -> InputRequestOption(
                    label = option["label"]?.jsonPrimitive?.contentOrNull.orEmpty(),
                    description = option["description"]?.jsonPrimitive?.contentOrNull.orEmpty(),
                    markdown = option["markdown"]?.jsonPrimitive?.contentOrNull,
                )
                else -> null
            }
        }
        return InputRequestQuestion(
            question = obj["question"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            header = obj["header"]?.jsonPrimitive?.contentOrNull,
            kind = kind,
            options = options,
            multiSelect = obj["multiSelect"]?.jsonPrimitive?.contentOrNull?.toBooleanStrictOrNull(),
        )
    }

    override fun serialize(encoder: Encoder, value: InputRequestQuestion) {
        val je = encoder as? JsonEncoder ?: error("InputRequestQuestion requires JSON")
        val map = buildMap<String, JsonElement> {
            put("question", JsonPrimitive(value.question))
            value.header?.let { put("header", JsonPrimitive(it)) }
            put("kind", je.json.encodeToJsonElement(InputRequestKind.serializer(), value.kind))
            value.options?.let {
                put("options", je.json.encodeToJsonElement(ListSerializer(InputRequestOption.serializer()), it))
            }
            value.multiSelect?.let { put("multiSelect", JsonPrimitive(it)) }
        }
        je.encodeJsonElement(JsonObject(map))
    }
}

File: apps/agor-rust-android/src/models/input_request.rs
use serde::{Deserialize, Deserializer, Serialize};
use serde_json::Value;
use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum InputRequestKind {
    FreeText,
    SingleChoice,
    MultiChoice,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum InputRequestStatus {
    Pending,
    Answered,
    Cancelled,
    TimedOut,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InputRequestOption {
    pub label: String,
    pub description: String,
    pub markdown: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct InputRequestQuestion {
    pub question: String,
    pub header: Option<String>,
    pub kind: InputRequestKind,
    pub options: Option<Vec<InputRequestOption>>,
    pub multi_select: Option<bool>,
}

impl Serialize for InputRequestQuestion {
    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        use serde::ser::SerializeMap;
        let mut map = serializer.serialize_map(None)?;
        map.serialize_entry("question", &self.question)?;
        if let Some(h) = &self.header {
            map.serialize_entry("header", h)?;
        }
        map.serialize_entry("kind", &self.kind)?;
        if let Some(o) = &self.options {
            map.serialize_entry("options", o)?;
        }
        if let Some(m) = &self.multi_select {
            map.serialize_entry("multiSelect", m)?;
        }
        map.end()
    }
}

impl<'de> Deserialize<'de> for InputRequestQuestion {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let v = Value::deserialize(deserializer)?;
        let obj = v.as_object().ok_or_else(|| serde::de::Error::custom("expected object"))?;

        let multi_select = obj
            .get("multiSelect")
            .or_else(|| obj.get("multi_select"))
            .and_then(|v| v.as_bool());

        let kind = if let Some(k) = obj.get("kind") {
            serde_json::from_value(k.clone()).unwrap_or(InputRequestKind::FreeText)
        } else if multi_select.unwrap_or(false) {
            InputRequestKind::MultiChoice
        } else if obj.get("options").map_or(false, |o| o.is_array()) {
            InputRequestKind::SingleChoice
        } else {
            InputRequestKind::FreeText
        };

        Ok(InputRequestQuestion {
            question: obj
                .get("question")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            header: obj.get("header").and_then(|v| v.as_str()).map(String::from),
            kind,
            options: obj
                .get("options")
                .and_then(|v| serde_json::from_value(v.clone()).ok()),
            multi_select,
        })
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InputRequestContent {
    pub request_id: String,
    pub legacy_input_request_id: Option<String>,
    pub task_id: Option<String>,
    pub questions: Vec<InputRequestQuestion>,
    pub status: InputRequestStatus,
    pub answers: Option<HashMap<String, String>>,
    pub answered_at: Option<String>,
    pub requested_at: Option<String>,
    pub context: Option<String>,
}

impl InputRequestContent {
    pub fn input_request_id(&self) -> &str {
        self.legacy_input_request_id
            .as_deref()
            .unwrap_or(&self.request_id)
    }
}
